crates/biome_js_analyze/src/semantic_analyzers/correctness/no_unused_variables.rs

```diff
@@ -273,45 +257,62 @@ impl Rule for NoUnusedVariables {
         if all_references.count() == 0 {
             Some(suggestion)
         } else {
-            // We need to check if all uses of this binding are somehow recursive
-
-            let function_declaration_scope = binding
-                .parent::<JsFunctionDeclaration>()
-                .map(|declaration| declaration.scope(model));
-
-            let declarator = binding.parent::<JsVariableDeclarator>();
-
-            let mut references_outside = 0;
-            for r in binding.all_references(model) {
-                let reference_scope = r.scope();
-
-                // If this binding is a function, and all its references are "inside" this
-                // function, we can safely say that this function is not used
-                if function_declaration_scope
-                    .as_ref()
-                    .map(|s| s.is_ancestor_of(&reference_scope))
-                    .unwrap_or(false)
+            // We need to check if all uses of this binding are somehow recursive or unused
+            let declaration = binding.declaration()?;
+            let declaration = declaration.syntax();
+            for reference in binding.all_references(model) {
+                let ref_parent = reference.syntax().parent()?;
+                if reference.is_write() {
+                    // Skip self assignment such as `a += 1` and `a++`.
+                    // Ensure that the assignment is not used in an used expression.
+                    let is_statement_like = ref_parent
+                        .ancestors()
+                        .find(|x| {
+                            JsAssignmentExpression::can_cast(x.kind())
+                                || JsPostUpdateExpression::can_cast(x.kind())
+                                || JsPreUpdateExpression::can_cast(x.kind())
+                        })
+                        .and_then(|x| is_unused_expression(&x).ok())
+                        .unwrap_or(false);
+                    if is_statement_like {
+                        continue;
+                    }
+                } else if JsIdentifierExpression::can_cast(ref_parent.kind())
+                    && is_unused_expression(&ref_parent).ok()?
                 {
+                    // The reference is in an unused expression
                     continue;
                 }
-
-                // Another possibility is if all its references are "inside" the same declaration
-                if let Some(declarator) = declarator.as_ref() {
-                    let node = declarator.syntax();
-                    if r.syntax().ancestors().any(|n| n == *node) {
-                        continue;
+                let mut is_inside = false;
+                let mut is_unused = true;
+                for ref ancestor in ref_parent.ancestors() {
```

Comment:
I am a bit hesitant about this loop inside a loop. The first loop could be a `filter_map` where we extract the `ref_parent` we need, the second could be a loop over the iterator emitted by `filter_map`.

Reply:
What your concern about the use of loops? Style? Readability?

Reply:
I wonder if we can use more iterator functions instead, they are more readable and we avoid middle mutable variables.

Reply:
I disagree on the readability part. However, I accept that the codebase is more functional oriented. I will try refactoring in a more functional style.